---
title: |
  FASTQ report    
  
output:
 html_document:
  highlight: tango
  theme: cosmo
  toc: no
  css: custom.css
params:
 fastq_dir:
  label: "Path to folder with fastq files (required, absolute path or relative to current folder)"
  value: "testdata"
  input: text
 fastq_pattern:
  label: "Regex to capture fastq files (and obtain sample names)"
  value: "fast(q|q.gz)$"
 subsample:
   label: "Subsample a fraction of the records to speed up calculations"
   value: 1
 platform:
  label: "Sequencing platform"
  choices: ["Illumina", "Nanopore", "PacBio"]
  value: "Nanopore"
  input: select
---
Generated with the [rfaster2](https://github.com/angelovangel/rfaster2) package (v`r packageVersion('rfaster2')`) on `r format.Date(Sys.time())`.   

***

Platform: <span class="badge"> `r params$platform` </span>    

```{r setup, include=FALSE}
knitr::opts_chunk$set(include = FALSE, 
                      echo = FALSE, 
                      warning = FALSE, 
                      cache = FALSE)

fastqdir <- normalizePath(params$fastq_dir)
fastqfiles <- list.files(fastqdir, pattern = params$fastq_pattern, full.names = TRUE, recursive = T)

if(length(fastqfiles) < 1) { 
  stop(paste0('No fastq files found: ', fastqfiles))
  }

names(fastqfiles) <- basename(fastqfiles)

```


```{r faster_table, echo = F, warning = F, message = F}

stats <- lapply(fastqfiles, rfaster2::fq_summary) %>% dplyr::bind_rows()

# get some summary data
num_files <- length(fastqfiles)
num_reads <- stats %>% dplyr::summarise(seqs = sum(reads)) %>% as.numeric()
num_reads_text <- label_number(scale_cut = cut_si(""), accuracy = 1)(num_reads)

num_bases <- stats %>% dplyr::summarise(bases = sum(bases)) %>% as.numeric()
num_bases_text <- label_number(scale_cut = cut_si(""), accuracy = 1)(num_bases)

```

***

### FASTQ read quality metrics
The report contains 
<blockquote>  
**`r num_files`** fastq files  
**`r num_reads_text`** reads  
**`r num_bases_text`** bases    
</blockquote>

```{r table1, include=TRUE}

kableExtra::kbl(
  stats, 
  digits = 2, caption = 'Table 1. FASTQ files statistics', 
  format.args = list(big.mark = ",")
  ) %>%
  # kable_styling(
  #   full_width = T, fixed_thead = TRUE, bootstrap_options = c("responsive")
  # ) %>%
  kable_minimal(lightable_options = 'hover')


```

***


```{r table2, include=TRUE, message=FALSE}
# table with gc and qscore histogram sparklines
sparkline(0) # load dependencies

# get nth based on subsample
thisnth = floor(1/params$subsample)

q_scores_densities <- lapply(fastqfiles, qscore_density, phred = T, nth = thisnth)
len_densities <- lapply(fastqfiles, len_density)
gc_densities <- lapply(fastqfiles, gc_density, nth = thisnth)
kmers <- lapply(fastqfiles, get_kmers, kmer = 3, nth = thisnth)

df <- data.frame(
	file = basename(fastqfiles),
	len_dist = sapply(len_densities, spark_len),
	gc_content_dist = sapply(gc_densities, spark_gc),
	q_score_dist = sapply(q_scores_densities, spark_phred),
	kmer_counts = sapply(kmers, spark_kmers), 
	row.names = NULL
  )
table2_caption <- paste0(
  "Table 2. Density distributions of length, GC-content and 'mean' q-score. 
  The q-scores 10, 20, 30, 40 are in red. The data has been subsampled  (fraction = ", round(params$subsample, 2), ")"
  )

df %>%
	#dplyr::arrange(file) %>%
	kableExtra::kbl(escape = F, caption = table2_caption) %>%
	#kable_styling(fixed_thead = TRUE, bootstrap_options = c("responsive"))
  kable_minimal()
  
```


