---
title: |
  FASTQ report    
  
output:
 html_document:
  highlight: tango
  theme: cosmo
  toc: no
  css: custom.css
params:
 fastq_dir:
  label: "Path to folder with fastq files (required, absolute path or relative to current folder)"
  value: "testdata"
  input: text
 fastq_pattern:
  label: "Regex to capture fastq files (and obtain sample names)"
  value: "fast(q|q.gz)$"
 platform:
  label: "Sequencing platform"
  choices: ["Illumina", "Nanopore", "PacBio"]
  value: "Nanopore"
  input: select
---
Generated with the [rfaster2](https://github.com/angelovangel/rfaster2) package on `r format.Date(Sys.time())`.   

***

Platform: <span class="badge"> `r params$platform` </span>    

```{r setup, include=FALSE}
knitr::opts_chunk$set(include = FALSE, 
                      echo = FALSE, 
                      warning = FALSE, 
                      cache = FALSE)

fastqdir <- normalizePath(params$fastq_dir)
fastqfiles <- list.files(fastqdir, pattern = params$fastq_pattern, full.names = TRUE, recursive = T)

if(length(fastqfiles) < 1) { 
  stop(paste0('No fastq files found: ', fastqfiles))
  }

names(fastqfiles) <- basename(fastqfiles)

```


```{r faster_table, echo = F, warning = F, message = F}

stats <- lapply(fastqfiles, rfaster2::fq_summary) %>% dplyr::bind_rows()

# get some summary data
num_files <- length(fastqfiles)
num_reads <- stats %>% dplyr::summarise(seqs = sum(reads)) %>% as.numeric()
num_reads_text <- label_number(scale_cut = cut_si(""), accuracy = 1)(num_reads)

num_bases <- stats %>% dplyr::summarise(bases = sum(bases)) %>% as.numeric()
num_bases_text <- label_number(scale_cut = cut_si(""), accuracy = 1)(num_bases)

```

***

### Number of reads and read quality metrics
The report contains 
<blockquote>  
**`r num_files`** fastq files  
**`r num_reads_text`** reads  
**`r num_bases_text`** bases    
</blockquote>

```{r table1, include=TRUE}

knitr::kable(stats, digits = 2, caption = 'Table 1. FASTQ files statistics', format.args = list(big.mark = ","))

```

***


```{r table2, include=TRUE, message=FALSE}
# table with gc and qscore histogram sparklines
sparkline(0) # load dependencies

spk_tool <- function(label, x, values) {
   htmlwidgets::JS(
     sprintf(
 		"function(sparkline, options, field){ return %s[field[0].offset]; }",
     jsonlite::toJSON(paste(label, x, ":",values, sep = " "))
     )
   )
}

qscore_density <- function(file) {
  
  qscores <- rfaster2::fq_quals(file, phred = T)
    # actually use density() here, not hist(). It returns a density list object with x and y, x is fixed from 1 to 50
    density_obj <- density(qscores, from = 1, to = 60, n = 60, na.rm = TRUE) # n is the number of equally spaced points at which the density is to be estimated.
    density_obj
}

spark_phred <- function(phred_density_obj) {
	#spk_chr(paste( round(phred_density_obj$x, digits = 2), ":", phred_density_obj$y, sep = ""),
  fillcolor <- "#5D6D7E"
	sparkline::spk_chr(round(phred_density_obj$y, digits = 2), 
					type = "bar",
					 # to highlight q-value of 30, only array (60 elements) seems to work, don't know how to pass range map here
					colorMap = c(
					  rep(fillcolor, 9), "red", 
					  rep(fillcolor, 9), "red", 
					  rep(fillcolor, 9), "red", 
					  rep(fillcolor, 9), "red", 
					  rep(fillcolor, 19)
					  ),
					width = 140, height = 40,
					tooltipFormatter = spk_tool("qscore ",phred_density_obj$x, round(phred_density_obj$y, 2))
					)
}

q_scores_density <- lapply(fastqfiles, qscore_density)

df <- data.frame(
	file = basename(fastqfiles),
	#gc_content_dist = sapply(gc_density, spark_gc),
	q_score_dist = sapply(q_scores_density, spark_phred),
	#len_dist = sapply(len_density, spark_len),
	#kmer_counts = sapply(kmers_tbl_list, spark_kmers), 
	row.names = NULL
  )
table2_caption <- "Table 2. Density distributions of length, GC-content and 'mean' q-score. The q-scores 10, 20, 30, 40 are in red"
df %>%
	#dplyr::arrange(file) %>%
	kableExtra::kbl(escape = F, caption = table2_caption) %>%
	kable_styling(fixed_thead = TRUE, bootstrap_options = c("responsive"))
  
```


