# Generated by extendr: Do not edit by hand

# nolint start

#
# This file was created with the following call:
#   .Call("wrap__make_rfaster2_wrappers", use_symbols = TRUE, package_name = "rfaster2")

#' @usage NULL
#' @useDynLib rfaster2, .registration = TRUE
NULL

#' FASTQ read lengths
#' 
#' Obtain read lengths from a fastq/fastq.gz file 
#' 
#' @param infile Path to fastq/fastq.gz file
#' @return Numeric vector with fastq records lengths
#' @export
fq_lengths <- function(infile) .Call(wrap__fq_lengths, infile)

#' FASTQ GC contents
#' 
#' Obtain per read GC contents from a fastq/fastq.gz file
#' 
#' @param infile Path to fastq/fastq.gz file
#' @param nth calculate gc for every nth record - used to subsample big fastq files
#' @return Numeric vector with GC content (fraction) per record
#' @export
fq_gc <- function(infile, nth) .Call(wrap__fq_gc, infile, nth)

#' FASTQ 'mean' quality scores
#' 
#' Obtain'mean' fastq read quality values. Mean Phred scores are calculated by converting to probabiliy, calculating mean quality, then converting back to Phred.
#' 
#' @param infile Path to fastq/fastq.gz file
#' @param phred Logical, report Phred score (error probability otherwise)
#' @param nth calculate qual for every nth record - used to subsample big fastq files
#' @return Numeric vector with 'mean' quality score per read
#' @export
fq_quals <- function(infile, phred, nth) .Call(wrap__fq_quals, infile, phred, nth)

#' FASTQ number of reads
#' 
#' Obtain the number of reads in a fastq/fastq.gz file
#' 
#' @param infile Path to fastq/fastq.gz file
#' @return Numeric 
#' @export
fq_reads <- function(infile) .Call(wrap__fq_reads, infile)

#' FASTQ number of bases
#' 
#' Obtain the number of bases in a fastq/fastq.gz file
#' 
#' @param infile Path to fastq/fastq.gz file
#' @return Numeric
#' @export
fq_bases <- function(infile) .Call(wrap__fq_bases, infile)

#' FASTQ k-mer counts
#' 
#' Get k-mer counts in a fastq/fastq.gz file
#' 
#' @param infile Path to fastq/fastq.gz file
#' @param k kmer length
#' @param nth calculate kmer for every nth record - used to subsample big fastq files
#' @return Numeric vector of k-mer counts (sorted by k-mer key)
#' @export
fq_kmers <- function(infile, k, nth) .Call(wrap__fq_kmers, infile, k, nth)

#' Return vector with summary values for the fastq file, this is not exported in the R package namespace
rust_summary <- function(infile) .Call(wrap__rust_summary, infile)


# nolint end
